### DISCUSSION TITLE
Is there a way to simply tell which channels are in use in a midi file?

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#190] [Q&A] [answered]
Is there a way to simply tell which channels are in use in a midi file for both the `SingleTrack` and `MultiTrack` types?
If I'm not wrong, a MIDI file contains only track-related metadata, but not channel.
So I think I need to iterate over all the timed events and see all the channel numbers used.
But also I think there should be a better way, so I'm asking.

Another question is to make sure that what I understood is correct.
Modifying objects returned by the getters such as `GetTrackChunks()`, `GetChannels()`, `GetNotes()` won't be saved when I call `MidiFile.Write()` because it's only possible via managers, or the special methods like `RemoveNotes()` and `ProcessNotes()`, right?

**Comment by melanchall** [comment, marked as answer]
1. You can use [GetChannels](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Core.MidiFileUtilities.html#Melanchall_DryWetMidi_Core_MidiFileUtilities_GetChannels_Melanchall_DryWetMidi_Core_MidiFile_) method. But you're right, you need to iterate over all MIDI events in a file. Saving kind of metadata creates more troubles than profit. Because every file modification requires to update that metadata, and there are a lot of ways to change MIDI file and all those ways need to properly update cached metadata.
2. Absolutely correct. Except `GetTrackChunks`. This method returns the same chunks a MIDI file contains. So direct operations on these chunks (for example, `GetTrackChunks().First().Events.First().DeltaTime = 20;`) will go to result file.

**Reply by Jenix-Park** [reply]
Thank you very much!
I love this library so much.
The only downside is this supports C# only.
I wish other MIDI libs for other languages were like this good.

**Reply by Jenix-Park** [reply]
Does the same apply to the tempo map?
I mean, do I need to call `MidiFile.ManageTempoMap()` and then `TempoMap` property instead of directly accessing `MidiFile.GetTempoMap()`?

**Reply by melanchall** [reply]
Yes, you need to use manager class to change a tempo map. `GetTempoMap` just returns current tempo map as immutable object.
